backend/libs/watcher-operations.ts

```diff
@@ -0,0 +1,116 @@
+import { DocumentClient } from 'aws-sdk/clients/dynamodb';
+import {
+    Room,
+    UserVideoStatus,
+    VideoSyncStatus,
+    Watcher,
+} from '../../extension/src/types';
+import {
+    MediaEventType,
+    PlayerEvent,
+} from './../../extension/src/contentscript/player';
+import { marshallMap } from './dynamodb-utils';
+
+/**
+ * Indicate that any info was received from a given watcher websocket connection
+ */
+const assignWatcherHeartbeat = (watcher: Watcher): void => {
+    watcher.lastHeartbeat = new Date();
+};
+
+/**
+ * Indicate a watcher has buffered enough video from the expected video timestamp, and could schedule a sync start
+ */
+const markWatcherAsReady = (watcher: Watcher) => {
+    watcher.currentVideoStatus = UserVideoStatus.READY;
+};
+
+/* eslint-disable complexity */
+/**
+ * Compute the new watcher status from his incoming player media event and the current room status
+ */
+const assignWatcherStatus = (
+    watcher: Watcher,
+    playerEvent: PlayerEvent,
+    room: Room,
+): void => {
+    watcher.lastVideoTimestamp = new Date(playerEvent.currentTime);
+    switch (playerEvent.mediaEventType) {
+        // TODO : https://github.com/LesGars/WatchWithMe/issues/116
+        // Right now we cannot differenciate "play:readyForSyncStart, play:buffering, play:syncStart"
+        // let's assume the current MediaEventType.PLAY means MediaEventType.PLAY:readyForSyncStart
+        case MediaEventType.PLAY: {
+            /*
+          if the room is NOT in PLAYING mode,
+            the play event should be intercepted by the extension to wait for sync
+          if the room is in PLAYING mode,
+            the play event should not be intercepted by the extension
+          */
+            switch (room.videoStatus) {
+                case VideoSyncStatus.PAUSED: {
+                    markWatcherAsReady(watcher);
+                    break;
+                }
+                case VideoSyncStatus.WAITING: {
+                    markWatcherAsReady(watcher);
+                    break;
+                }
+                case VideoSyncStatus.PLAYING: {
+                    watcher.currentVideoStatus = UserVideoStatus.PLAYING;
```

Comment:
Shouldn't we get his timestamp and his localTime ?

Reply:
Yes, so far it was not even modelled in the watcher type, I'll work on in a future commit